source/s3_meta_request.c

```diff
@@ -0,0 +1,673 @@
+            PyObject_CallMethod(request_binding->self_py, "_on_progress", "(K)", request_binding->size_transferred);
+        if (!result) {
+            PyErr_WriteUnraisable(request_binding->self_py);
+        } else {
+            Py_DECREF(result);
+        }
+        request_binding->size_transferred = 0;
+    }
+    error = false;
+done:
+    PyGILState_Release(state);
+    if (error) {
+        return aws_raise_error(AWS_ERROR_CRT_CALLBACK_EXCEPTION);
+    } else {
+        return AWS_OP_SUCCESS;
+    }
+}
+
+/* If the request has not finished, it will keep the request alive, until the finish callback invoked. So, we don't
+ * need to clean anything from this call */
+static void s_s3_request_on_finish(
+    struct aws_s3_meta_request *meta_request,
+    const struct aws_s3_meta_request_result *meta_request_result,
+    void *user_data) {
+    (void)meta_request;
+    struct s3_meta_request_binding *request_binding = user_data;
+
+    /*************** GIL ACQUIRE ***************/
+    PyGILState_STATE state;
+    if (aws_py_gilstate_ensure(&state)) {
+        return; /* Python has shut down. Nothing matters anymore, but don't crash */
+    }
+
+    PyObject *header_list = NULL;
+    PyObject *result = NULL;
+
+    if (request_binding->size_transferred) {
+        /* report the remaining progress */
+        result =
+            PyObject_CallMethod(request_binding->self_py, "_on_progress", "(K)", request_binding->size_transferred);
+        if (!result) {
+            PyErr_WriteUnraisable(request_binding->self_py);
+        } else {
+            Py_DECREF(result);
+        }
+        request_binding->size_transferred = 0;
+    }
+    struct aws_byte_buf error_body;
+    AWS_ZERO_STRUCT(error_body);
+    /* Get the header and body of the error */
+    if (meta_request_result->error_response_headers) {
+        header_list = PyList_New(aws_http_headers_count(meta_request_result->error_response_headers));
+        if (!header_list) {
+            PyErr_WriteUnraisable(request_binding->self_py);
+            goto done;
+        }
+        if (s_get_py_headers(meta_request_result->error_response_headers, header_list)) {
+            PyErr_WriteUnraisable(request_binding->self_py);
+            goto done;
+        }
+    }
+    if (meta_request_result->error_response_body) {
+        error_body = *(meta_request_result->error_response_body);
+    }
+    result = PyObject_CallMethod(
+        request_binding->self_py,
+        "_on_finish",
+        "(iOy#)",
+        meta_request_result->error_code,
+        header_list ? header_list : Py_None,
+        (const char *)(error_body.buffer),
+        (Py_ssize_t)error_body.len);
+
+    if (result) {
+        Py_DECREF(result);
+    } else {
+        /* Callback might fail during application shutdown */
+        PyErr_WriteUnraisable(request_binding->self_py);
+    }
+done:
+    Py_XDECREF(header_list);
+    Py_CLEAR(request_binding->self_py);
+    Py_CLEAR(request_binding->http_message_py);
```

Comment:
you're might remove the http_message_py variable from existence anyway...

but if you keep it, this isn't the right place to clear it
it still needs to exist until the shutdown callback fires, right?
because if we cancel other threads might still be reading from it?

Reply:
Yep, removed!